components/scream/src/physics/p3/p3_functions_table_ice_impl.hpp

```diff
@@ -0,0 +1,292 @@
+#define P3_FUNCTIONS_TABLE_ICE_IMPL_HPP
+
+#include <fstream>
+
+#include "p3_functions.hpp"
+#include "p3_constants.hpp"
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 ice table functions. Clients should NOT #include
+ * this file, #include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+void Functions<S,D>
+::init_kokkos_ice_lookup_tables(view_itab_table& itab, view_itabcol_table& itabcol) {
+
+  using DeviceItab    = typename view_itab_table::non_const_type;
+  using DeviceItabcol = typename view_itabcol_table::non_const_type;
+
+  const auto itab_d    = DeviceItab("itab");
+  const auto itabcol_d = DeviceItabcol("itabcol");
+
+  const auto itab_h    = Kokkos::create_mirror_view(itab_d);
+  const auto itabcol_h = Kokkos::create_mirror_view(itabcol_d);
+
+  //
+  // read in ice microphysics table into host views
+  //
+
+  std::string filename = std::string(P3C::p3_lookup_base) + std::string(P3C::p3_version);
+
+  std::ifstream in(filename);
+
+  // read header
+  std::string version, version_val;
+  in >> version >> version_val;
+  scream_require_msg(version == "VERSION", "Bad " << filename << ", expected VERSION X.Y.Z header");
+  scream_require_msg(version_val == P3C::p3_version, "Bad " << filename << ", expected version " << P3C::p3_version << ", but got " << version_val);
+
+  // read tables
+  double dum_s; int dum_i; // dum_s needs to be double to stream correctly
+  for (int jj = 0; jj < P3C::densize; ++jj) {
+    for (int ii = 0; ii < P3C::rimsize; ++ii) {
+      for (int i = 0; i < P3C::isize; ++i) {
+        in >> dum_i >> dum_i;
+        int j_idx = 0;
+        for (int j = 0; j < 15; ++j) {
+          in >> dum_s;
+          if (j > 1 && j != 10) {
+            itab_h(jj, ii, i, j_idx++) = dum_s;
+          }
+        }
+      }
+
+      for (int i = 0; i < P3C::isize; ++i) {
+        for (int j = 0; j < P3C::rcollsize; ++j) {
+          in >> dum_i >> dum_i;
+          int k_idx = 0;
+          for (int k = 0; k < 6; ++k) {
+            in >> dum_s;
+            if (k == 3 || k == 4) {
+              itabcol_h(jj, ii, i, j, k_idx++) = std::log10(dum_s);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // deep copy to device
+  Kokkos::deep_copy(itab_d, itab_h);
+  Kokkos::deep_copy(itabcol_d, itabcol_h);
+  itab    = itab_d;
+  itabcol = itabcol_d;
+}
+
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::lookup_ice (const Smask& qiti_gt_small, const Spack& qitot, const Spack& nitot,
+              const Spack& qirim, const Spack& rhop, TableIce& t)
+{
+  // find index for qi (normalized ice mass mixing ratio = qitot/nitot)
+  //   dum1 = (log10(qitot)+16.)/0.70757  !orig
+  //   dum1 = (log10(qitot)+16.)*1.41328
+  // we are inverting this equation from the lookup table to solve for i:
+  // qitot/nitot=261.7**((i+10)*0.1)*1.e-18
+  // dum1 = (log10(qitot/nitot)+18.)/(0.1*log10(261.7))-10.
+
+  if (qiti_gt_small.any()) {
```

Comment:
The entire packs get set, so that covers all the slots where qiti_qt_small it true. I think this is OK as long as the pack slots where !qiti_gt_small are ignored by the broader algorithm, which I think is true. FPE is also safe since, when pksize==1, the any() prevents execution entirely if !qiti_gt_small.

Reply:
OK. As a matter of style and opinion only (so no need to change), if you have a function like this:
```
void foo () {
  if (something) {
    // body
  }
}
```
it can be nicer to write
```
void foo () {
  if ( ! something) return; // it's clear there's nothing to do in the !something case
  // body with 1 less indent
}
```